sycl-fusion/common/include/DynArray.h

```diff
@@ -0,0 +1,97 @@
+//==------------ DynArray.h - Non-STL replacement for std::array -----------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SYCL_FUSION_COMMON_DYNARRAY_H
+#define SYCL_FUSION_COMMON_DYNARRAY_H
+
+#include <algorithm>
+#include <cassert>
+
+namespace jit_compiler {
+
+///
+/// A fixed-size, dynamically-allocated array, with an interface that is a
+/// subset of `std::array`.
+template <typename T> class DynArray {
+public:
+  DynArray() {}
+
+  explicit DynArray(size_t Size) {
+    if (Size > 0)
+      init(Size);
+  }
+
+  ~DynArray() { deinit(); }
+
+  DynArray(const DynArray &Other) {
+    if (Other.Size > 0) {
+      init(Other.Size);
+      std::copy(Other.begin(), Other.end(), begin());
+    }
+  }
+
+  DynArray &operator=(const DynArray &Other) {
+    deinit();
+    if (Other.Size > 0) {
+      init(Other.Size);
+      std::copy(Other.begin(), Other.end(), begin());
+    }
+    return *this;
+  }
+
+  DynArray(DynArray &&Other) { swap(std::move(Other)); }
+
+  DynArray &operator=(DynArray &&Other) {
+    deinit();
+    swap(std::move(Other));
+    return *this;
+  }
+
+  size_t size() const { return Size; }
+  bool empty() const { return Size == 0; }
+
+  const T *begin() const { return Values; }
+  const T *end() const { return Values + Size; }
+  T *begin() { return Values; }
+  T *end() { return Values + Size; }
+
+  const T &operator[](int Idx) const { return Values[Idx]; }
+  T &operator[](int Idx) { return Values[Idx]; }
+
+private:
+  T *Values = nullptr;
+  size_t Size = 0;
+
+  void init(size_t NewSize) {
+    assert(!Values && Size == 0 && "Cannot init(..) twice");
+    assert(NewSize > 0 && "Cannot init(..) empty dynarray");
+    Values = new T[NewSize];
+    Size = NewSize;
+  }
+
+  void deinit() {
+    if (!Values && Size == 0) {
+      return;
+    }
+    assert(Values && Size > 0 && "Cannot deinit() corrupted dynarray");
+    delete[] Values;
+    Values = nullptr;
+    Size = 0;
+  }
+
+  void swap(DynArray &&Other) {
+    assert(!Values && Size == 0 && "Cannot swap(..) into initialized object");
+    Values = Other.Values, Other.Values = nullptr;
+    Size = Other.Size;
+    Other.Size = 0;
+  }
```

Comment:
Could be confused with conventional `swap`. Can we rename it, e.g., `move_from`?
```suggestion
  void move_from(DynArray &&Other) {
    assert(!Values && Size == 0 && "Cannot swap(..) into initialized object");
    Values = Other.Values;
    Other.Values = nullptr;
    Size = Other.Size;
    Other.Size = 0;
  }
```

Reply:
Going with `moveFrom` to appease the style checker.